Assets/_Scripts/CurveController.cs

```diff
@@ -271,21 +256,26 @@ private void UpdateFrameNumber()
         }
     }
 
-    private void UpdateALLMarkersRotation()
+    private void UpdateMarkersRotation(List<int> affectedMarkers)
```

Comment:
This is also another problem that will cause you to do some dirty workarounds to access the markers when you need them. You are passing the markers as integers (their index) here. This made you in other methods such as ShiftDeleteMarker, DeleteMarker, etc. to create new temporary lists (this is not a good performance practice to create new lists in a method that might be called a lot, for example, every frame) Instead you could store the marker's index inside the MarkerEntity and then pass the MarkerEntity to those events and return a list of MarkerEntities back and work with them. A trade-off here is that you have to always update the MarkerIndex inside MarkerEntity.

Reply:
You could also add a new flag to the MarkerEntity called "NeedsVisualUpdate" whenever something about the marker changes (for example rotation needs to be updated, or the position, etc) and then have a phase in every method that iterates over all the markers and updates those who have this flag on and then turns off the flag.